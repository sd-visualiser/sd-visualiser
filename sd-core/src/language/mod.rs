use std::{
    fmt::{Debug, Display},
    hash::Hash,
};

use derivative::Derivative;

use crate::{common::Matchable, hypergraph::traits::WithType, prettyprinter::PrettyPrint};

pub mod chil;
#[cfg(not(target_arch = "wasm32"))]
pub mod llvm_ir;
pub mod mlir;
pub mod sd_lang;

pub(crate) fn span_into_str(span: pest::Span) -> &str {
    span.as_str()
}

pub trait GetVar<V> {
    fn var(&self) -> &V;
    fn into_var(self) -> V;
}

impl<V> GetVar<V> for V {
    fn var(&self) -> &V {
        self
    }

    fn into_var(self) -> V {
        self
    }
}

pub trait Fresh {
    fn fresh(number: usize) -> Self;
}

/// `Display` should give the symbolic representation (e.g. "+").
/// `PrettyPrint` should give the textual representation (e.g. "plus").
pub trait Syntax:
    Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint
{
}
impl<T: Clone + Eq + Hash + Debug + Send + Sync + Display + Matchable + PrettyPrint> Syntax for T {}

pub enum CF<T: Language + ?Sized> {
    Return,
    Brs(Vec<T::BlockAddr>),
}

pub trait OpInfo<T: Language + ?Sized> {
    fn get_cf(&self) -> Option<CF<T>> {
        None
    }
    fn symbols_used(&self) -> impl Iterator<Item = T::Symbol> {
        std::iter::empty()
    }
    fn sym_name(&self) -> Option<T::Symbol> {
        None
    }
}

pub trait Language {
    type Op: Syntax + OpInfo<Self>;
    type Var: Syntax + Fresh + From<Self::Symbol> + WithType;
    type Addr: Syntax;
    type BlockAddr: Syntax;
    type VarDef: Syntax + GetVar<Self::Var>;
    type Symbol: Syntax + TryFrom<Self::Addr, Error: Debug>;
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = ""),
    Default(bound = "")
)]
pub struct Expr<T: Language + ?Sized> {
    pub binds: Vec<Bind<T>>,
    pub values: Vec<Value<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Bind<T: Language + ?Sized> {
    pub defs: Vec<T::VarDef>,
    pub value: Value<T>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub enum Value<T: Language + ?Sized> {
    Variable(T::Var),
    Thunk(Thunk<T>),
    Op { op: T::Op, args: Vec<Value<T>> },
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Thunk<T: Language + ?Sized> {
    pub addr: T::Addr,
    pub args: Vec<T::VarDef>,
    pub reqs: Vec<T::Var>,
    pub body: Expr<T>,
    pub blocks: Vec<Block<T>>,
}

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Eq(bound = ""),
    PartialEq(bound = ""),
    Hash(bound = ""),
    Debug(bound = "")
)]
pub struct Block<T: Language + ?Sized> {
    pub addr: T::BlockAddr,
    pub args: Vec<T::VarDef>,
    pub expr: Expr<T>,
}

// Conversions between languages

impl<T: Language> Expr<T> {
    pub fn into<U: Language>(self) -> Expr<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Expr {
            binds: self.binds.into_iter().map(Bind::into).collect(),
            values: self.values.into_iter().map(Value::into).collect(),
        }
    }
}

impl<T: Language> Bind<T> {
    pub fn into<U: Language>(self) -> Bind<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Bind {
            defs: self.defs.into_iter().map(Into::into).collect(),
            value: self.value.into(),
        }
    }
}

impl<T: Language> Value<T> {
    pub fn into<U: Language>(self) -> Value<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        match self {
            Self::Variable(var) => Value::Variable(var.into()),
            Self::Thunk(thunk) => Value::Thunk(thunk.into()),
            Self::Op { op, args } => Value::Op {
                op: op.into(),
                args: args.into_iter().map(Value::into).collect(),
            },
        }
    }
}

impl<T: Language> Thunk<T> {
    pub fn into<U: Language>(self) -> Thunk<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Thunk {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            reqs: self.reqs.into_iter().map(Into::into).collect(),
            body: self.body.into(),
            blocks: self.blocks.into_iter().map(Block::into).collect(),
        }
    }
}

impl<T: Language> Block<T> {
    pub fn into<U: Language>(self) -> Block<U>
    where
        U::Op: From<T::Op>,
        U::Var: From<T::Var>,
        U::Addr: From<T::Addr>,
        U::VarDef: From<T::VarDef>,
        U::BlockAddr: From<T::BlockAddr>,
    {
        Block {
            addr: self.addr.into(),
            args: self.args.into_iter().map(Into::into).collect(),
            expr: self.expr.into(),
        }
    }
}

#[cfg(test)]
pub(crate) mod tests {
    use std::{ffi::OsStr, path::Path};

    use serde::Serialize;

    use super::{
        Expr, Language,
        chil::tests::parse_chil,
        llvm_ir::tests::parse_llvm_ir,
        mlir::{
            self,
            internal::{TopLevelItem, tests::parse_mlir},
        },
        sd_lang::tests::parse_sd_lang,
    };
    use crate::{graph::SyntaxHypergraph, hypergraph::petgraph::to_pet, language::llvm_ir::LlvmIr};

    pub trait ExprTest {
        fn free_var_test(&self) -> Box<dyn std::fmt::Debug>;
        fn graph_test(&self, name: &str, lang: &str, sym_name_link: bool) -> anyhow::Result<()>;
    }

    impl<T: Language + 'static> ExprTest for Expr<T>
    where
        <T as Language>::Op: Serialize,
        <T as Language>::BlockAddr: Serialize,
        <T as Language>::Var: Serialize,
        <T as Language>::VarDef: Serialize,
    {
        fn free_var_test(&self) -> Box<dyn std::fmt::Debug> {
            Box::new(self.free_vars(false))
        }

        fn graph_test(&self, name: &str, lang: &str, sym_name_link: bool) -> anyhow::Result<()> {
            let graph: SyntaxHypergraph<T> = self.to_graph(sym_name_link)?;
            let name = if sym_name_link {
                format!("hypergraph_with_sym_{name}.{lang}")
            } else {
                format!("hypergraph_{name}.{lang}")
            };
            insta::assert_ron_snapshot!(name, to_pet(&graph));
            Ok(())
        }
    }

    impl ExprTest for crate::language::llvm_ir::Expr {
        fn free_var_test(&self) -> Box<dyn std::fmt::Debug> {
            Box::new(self.free_vars(false))
        }

        fn graph_test(&self, name: &str, lang: &str, sym_name_link: bool) -> anyhow::Result<()> {
            let graph: SyntaxHypergraph<LlvmIr> = self.to_graph(sym_name_link)?;
            let name = if sym_name_link {
                format!("hypergraph_with_sym_{name}_debug.{lang}")
            } else {
                format!("hypergraph_{name}_debug.{lang}")
            };
            // cannot serialise LLVM IR expressions, so use Debug instead
            insta::assert_debug_snapshot!(name, to_pet(&graph));
            Ok(())
        }
    }

    pub fn parse(raw_path: &str) -> (&str, &str, Box<dyn ExprTest>) {
        let path = Path::new(raw_path);
        match path.extension() {
            Some(ext) if ext == OsStr::new("sd") => {
                let (name, expr) = parse_sd_lang(raw_path);
                ("sd", name, Box::new(expr))
            }
            Some(ext) if ext == OsStr::new("chil") => {
                let (name, expr) = parse_chil(raw_path);
                ("chil", name, Box::new(expr))
            }
            Some(ext) if ext == OsStr::new("ll") => {
                let (name, expr) = parse_llvm_ir(raw_path);
                ("llvm-ir", name, Box::new(expr))
            }
            Some(ext) if ext == OsStr::new("mlir") => {
                let (name, items) = parse_mlir(raw_path);
                let ops: Vec<mlir::internal::Operation> = items
                    .into_iter()
                    .filter_map(|x| match x {
                        TopLevelItem::Operation(y) => Some(y),
                        TopLevelItem::Other(_) => None,
                    })
                    .collect();
                let expr = mlir::Expr::from(ops);
                ("mlir", name, Box::new(expr))
            }
            _ => unreachable!(),
        }
    }
}
