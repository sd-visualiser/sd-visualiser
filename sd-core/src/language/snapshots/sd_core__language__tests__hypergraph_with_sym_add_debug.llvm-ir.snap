---
source: sd-core/src/language/mod.rs
expression: to_pet(&graph)
---
Graph {
    Ty: "Directed",
    node_count: 4,
    edge_count: 3,
    edges: (1, 0), (0, 1), (2, 3),
    node weights: {
        0: Thunk(
            Graph {
                Ty: "Directed",
                node_count: 10,
                edge_count: 11,
                edges: (2, 3), (4, 3), (3, 5), (0, 6), (2, 6), (1, 6), (7, 6), (6, 8), (1, 9), (9, 4), (9, 7),
                node weights: {
                    0: Edge(
                        FreeVar(
                            Symbol {
                                symbol: "add2",
                            },
                        ),
                    ),
                    1: Edge(
                        BoundVar(
                            Var {
                                name: Name(
                                    "a",
                                ),
                            },
                        ),
                    ),
                    2: Edge(
                        BoundVar(
                            Var {
                                name: Name(
                                    "b",
                                ),
                            },
                        ),
                    ),
                    3: Thunk(
                        Graph {
                            Ty: "Directed",
                            node_count: 3,
                            edge_count: 2,
                            edges: (0, 1), (1, 2),
                            node weights: {
                                0: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "b",
                                            ),
                                        },
                                    ),
                                ),
                                1: Operation(
                                    Terminator(
                                        Ret(
                                            Ret {
                                                return_operand: Some(
                                                    LocalOperand {
                                                        name: Name(
                                                            "b",
                                                        ),
                                                        ty: TypeRef(
                                                            IntegerType {
                                                                bits: 32,
                                                            },
                                                        ),
                                                    },
                                                ),
                                                debugloc: None,
                                            },
                                        ),
                                    ),
                                ),
                                2: Edge(
                                    CF(
                                        None,
                                    ),
                                ),
                            },
                            edge weights: {
                                0: 0,
                                1: 0,
                            },
                        },
                    ),
                    4: Edge(
                        CF(
                            Some(
                                Name(
                                    "done",
                                ),
                            ),
                        ),
                    ),
                    5: Edge(
                        CF(
                            None,
                        ),
                    ),
                    6: Thunk(
                        Graph {
                            Ty: "Directed",
                            node_count: 17,
                            edge_count: 16,
                            edges: (4, 3), (3, 5), (7, 6), (8, 6), (9, 6), (6, 4), (0, 10), (10, 7), (1, 11), (12, 11), (11, 9), (13, 12), (2, 14), (15, 14), (14, 8), (16, 15),
                            node weights: {
                                0: Edge(
                                    FreeVar(
                                        Symbol {
                                            symbol: "add2",
                                        },
                                    ),
                                ),
                                1: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "b",
                                            ),
                                        },
                                    ),
                                ),
                                2: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "a",
                                            ),
                                        },
                                    ),
                                ),
                                3: Operation(
                                    Terminator(
                                        Ret(
                                            Ret {
                                                return_operand: Some(
                                                    LocalOperand {
                                                        name: Name(
                                                            "tmp4",
                                                        ),
                                                        ty: TypeRef(
                                                            IntegerType {
                                                                bits: 32,
                                                            },
                                                        ),
                                                    },
                                                ),
                                                debugloc: None,
                                            },
                                        ),
                                    ),
                                ),
                                4: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "tmp4",
                                            ),
                                        },
                                    ),
                                ),
                                5: Edge(
                                    CF(
                                        None,
                                    ),
                                ),
                                6: Operation(
                                    Instruction(
                                        Call(
                                            Call {
                                                function: Right(
                                                    ConstantOperand(
                                                        ConstantRef(
                                                            GlobalReference {
                                                                name: Name(
                                                                    "add2",
                                                                ),
                                                                ty: TypeRef(
                                                                    FuncType {
                                                                        result_type: TypeRef(
                                                                            IntegerType {
                                                                                bits: 32,
                                                                            },
                                                                        ),
                                                                        param_types: [
                                                                            TypeRef(
                                                                                IntegerType {
                                                                                    bits: 32,
                                                                                },
                                                                            ),
                                                                            TypeRef(
                                                                                IntegerType {
                                                                                    bits: 32,
                                                                                },
                                                                            ),
                                                                        ],
                                                                        is_var_arg: false,
                                                                    },
                                                                ),
                                                            },
                                                        ),
                                                    ),
                                                ),
                                                function_ty: TypeRef(
                                                    FuncType {
                                                        result_type: TypeRef(
                                                            IntegerType {
                                                                bits: 32,
                                                            },
                                                        ),
                                                        param_types: [
                                                            TypeRef(
                                                                IntegerType {
                                                                    bits: 32,
                                                                },
                                                            ),
                                                            TypeRef(
                                                                IntegerType {
                                                                    bits: 32,
                                                                },
                                                            ),
                                                        ],
                                                        is_var_arg: false,
                                                    },
                                                ),
                                                arguments: [
                                                    (
                                                        LocalOperand {
                                                            name: Name(
                                                                "tmp2",
                                                            ),
                                                            ty: TypeRef(
                                                                IntegerType {
                                                                    bits: 32,
                                                                },
                                                            ),
                                                        },
                                                        [],
                                                    ),
                                                    (
                                                        LocalOperand {
                                                            name: Name(
                                                                "tmp3",
                                                            ),
                                                            ty: TypeRef(
                                                                IntegerType {
                                                                    bits: 32,
                                                                },
                                                            ),
                                                        },
                                                        [],
                                                    ),
                                                ],
                                                return_attributes: [],
                                                dest: Some(
                                                    Name(
                                                        "tmp4",
                                                    ),
                                                ),
                                                function_attributes: [],
                                                is_tail_call: false,
                                                calling_convention: C,
                                                debugloc: None,
                                            },
                                        ),
                                    ),
                                ),
                                7: Edge(
                                    Nil,
                                ),
                                8: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "tmp2",
                                            ),
                                        },
                                    ),
                                ),
                                9: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "tmp3",
                                            ),
                                        },
                                    ),
                                ),
                                10: Operation(
                                    Constant(
                                        ConstantRef(
                                            GlobalReference {
                                                name: Name(
                                                    "add2",
                                                ),
                                                ty: TypeRef(
                                                    FuncType {
                                                        result_type: TypeRef(
                                                            IntegerType {
                                                                bits: 32,
                                                            },
                                                        ),
                                                        param_types: [
                                                            TypeRef(
                                                                IntegerType {
                                                                    bits: 32,
                                                                },
                                                            ),
                                                            TypeRef(
                                                                IntegerType {
                                                                    bits: 32,
                                                                },
                                                            ),
                                                        ],
                                                        is_var_arg: false,
                                                    },
                                                ),
                                            },
                                        ),
                                    ),
                                ),
                                11: Operation(
                                    Instruction(
                                        Add(
                                            Add {
                                                operand0: LocalOperand {
                                                    name: Name(
                                                        "b",
                                                    ),
                                                    ty: TypeRef(
                                                        IntegerType {
                                                            bits: 32,
                                                        },
                                                    ),
                                                },
                                                operand1: ConstantOperand(
                                                    ConstantRef(
                                                        Int {
                                                            bits: 32,
                                                            value: 1,
                                                        },
                                                    ),
                                                ),
                                                dest: Name(
                                                    "tmp3",
                                                ),
                                                nuw: false,
                                                nsw: false,
                                                debugloc: None,
                                            },
                                        ),
                                    ),
                                ),
                                12: Edge(
                                    Nil,
                                ),
                                13: Operation(
                                    Constant(
                                        ConstantRef(
                                            Int {
                                                bits: 32,
                                                value: 1,
                                            },
                                        ),
                                    ),
                                ),
                                14: Operation(
                                    Instruction(
                                        Sub(
                                            Sub {
                                                operand0: LocalOperand {
                                                    name: Name(
                                                        "a",
                                                    ),
                                                    ty: TypeRef(
                                                        IntegerType {
                                                            bits: 32,
                                                        },
                                                    ),
                                                },
                                                operand1: ConstantOperand(
                                                    ConstantRef(
                                                        Int {
                                                            bits: 32,
                                                            value: 1,
                                                        },
                                                    ),
                                                ),
                                                dest: Name(
                                                    "tmp2",
                                                ),
                                                nuw: false,
                                                nsw: false,
                                                debugloc: None,
                                            },
                                        ),
                                    ),
                                ),
                                15: Edge(
                                    Nil,
                                ),
                                16: Operation(
                                    Constant(
                                        ConstantRef(
                                            Int {
                                                bits: 32,
                                                value: 1,
                                            },
                                        ),
                                    ),
                                ),
                            },
                            edge weights: {
                                0: 0,
                                1: 0,
                                2: 0,
                                3: 1,
                                4: 2,
                                5: 0,
                                6: 0,
                                7: 0,
                                8: 0,
                                9: 1,
                                10: 0,
                                11: 0,
                                12: 0,
                                13: 1,
                                14: 0,
                                15: 0,
                            },
                        },
                    ),
                    7: Edge(
                        CF(
                            Some(
                                Name(
                                    "recurse",
                                ),
                            ),
                        ),
                    ),
                    8: Edge(
                        CF(
                            None,
                        ),
                    ),
                    9: Thunk(
                        Graph {
                            Ty: "Directed",
                            node_count: 10,
                            edge_count: 9,
                            edges: (2, 1), (1, 3), (1, 4), (6, 5), (0, 5), (7, 5), (5, 2), (8, 7), (9, 6),
                            node weights: {
                                0: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "a",
                                            ),
                                        },
                                    ),
                                ),
                                1: Operation(
                                    Terminator(
                                        CondBr(
                                            CondBr {
                                                condition: LocalOperand {
                                                    name: Name(
                                                        "tmp1",
                                                    ),
                                                    ty: TypeRef(
                                                        IntegerType {
                                                            bits: 1,
                                                        },
                                                    ),
                                                },
                                                true_dest: Name(
                                                    "done",
                                                ),
                                                false_dest: Name(
                                                    "recurse",
                                                ),
                                                debugloc: None,
                                            },
                                        ),
                                    ),
                                ),
                                2: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "tmp1",
                                            ),
                                        },
                                    ),
                                ),
                                3: Edge(
                                    CF(
                                        Some(
                                            Name(
                                                "done",
                                            ),
                                        ),
                                    ),
                                ),
                                4: Edge(
                                    CF(
                                        Some(
                                            Name(
                                                "recurse",
                                            ),
                                        ),
                                    ),
                                ),
                                5: Operation(
                                    Instruction(
                                        ICmp(
                                            ICmp {
                                                predicate: EQ,
                                                operand0: LocalOperand {
                                                    name: Name(
                                                        "a",
                                                    ),
                                                    ty: TypeRef(
                                                        IntegerType {
                                                            bits: 32,
                                                        },
                                                    ),
                                                },
                                                operand1: ConstantOperand(
                                                    ConstantRef(
                                                        Int {
                                                            bits: 32,
                                                            value: 0,
                                                        },
                                                    ),
                                                ),
                                                dest: Name(
                                                    "tmp1",
                                                ),
                                                debugloc: None,
                                            },
                                        ),
                                    ),
                                ),
                                6: Edge(
                                    Nil,
                                ),
                                7: Edge(
                                    Nil,
                                ),
                                8: Operation(
                                    Constant(
                                        ConstantRef(
                                            Int {
                                                bits: 32,
                                                value: 0,
                                            },
                                        ),
                                    ),
                                ),
                                9: Operation(
                                    IntPredicate(
                                        EQ,
                                    ),
                                ),
                            },
                            edge weights: {
                                0: 0,
                                1: 0,
                                2: 1,
                                3: 0,
                                4: 1,
                                5: 2,
                                6: 0,
                                7: 0,
                                8: 0,
                            },
                        },
                    ),
                },
                edge weights: {
                    0: 0,
                    1: 1,
                    2: 0,
                    3: 0,
                    4: 1,
                    5: 2,
                    6: 3,
                    7: 0,
                    8: 0,
                    9: 0,
                    10: 1,
                },
            },
        ),
        1: Edge(
            FreeVar(
                Symbol {
                    symbol: "add2",
                },
            ),
        ),
        2: Thunk(
            Graph {
                Ty: "Directed",
                node_count: 4,
                edge_count: 3,
                edges: (0, 2), (1, 2), (2, 3),
                node weights: {
                    0: Edge(
                        BoundVar(
                            Var {
                                name: Name(
                                    "a",
                                ),
                            },
                        ),
                    ),
                    1: Edge(
                        BoundVar(
                            Var {
                                name: Name(
                                    "b",
                                ),
                            },
                        ),
                    ),
                    2: Thunk(
                        Graph {
                            Ty: "Directed",
                            node_count: 6,
                            edge_count: 5,
                            edges: (3, 2), (2, 4), (0, 5), (1, 5), (5, 3),
                            node weights: {
                                0: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "a",
                                            ),
                                        },
                                    ),
                                ),
                                1: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "b",
                                            ),
                                        },
                                    ),
                                ),
                                2: Operation(
                                    Terminator(
                                        Ret(
                                            Ret {
                                                return_operand: Some(
                                                    LocalOperand {
                                                        name: Name(
                                                            "tmp1",
                                                        ),
                                                        ty: TypeRef(
                                                            IntegerType {
                                                                bits: 32,
                                                            },
                                                        ),
                                                    },
                                                ),
                                                debugloc: None,
                                            },
                                        ),
                                    ),
                                ),
                                3: Edge(
                                    BoundVar(
                                        Var {
                                            name: Name(
                                                "tmp1",
                                            ),
                                        },
                                    ),
                                ),
                                4: Edge(
                                    CF(
                                        None,
                                    ),
                                ),
                                5: Operation(
                                    Instruction(
                                        Add(
                                            Add {
                                                operand0: LocalOperand {
                                                    name: Name(
                                                        "a",
                                                    ),
                                                    ty: TypeRef(
                                                        IntegerType {
                                                            bits: 32,
                                                        },
                                                    ),
                                                },
                                                operand1: LocalOperand {
                                                    name: Name(
                                                        "b",
                                                    ),
                                                    ty: TypeRef(
                                                        IntegerType {
                                                            bits: 32,
                                                        },
                                                    ),
                                                },
                                                dest: Name(
                                                    "tmp1",
                                                ),
                                                nuw: false,
                                                nsw: false,
                                                debugloc: None,
                                            },
                                        ),
                                    ),
                                ),
                            },
                            edge weights: {
                                0: 0,
                                1: 0,
                                2: 0,
                                3: 1,
                                4: 0,
                            },
                        },
                    ),
                    3: Edge(
                        CF(
                            None,
                        ),
                    ),
                },
                edge weights: {
                    0: 0,
                    1: 1,
                    2: 0,
                },
            },
        ),
        3: Edge(
            FreeVar(
                Symbol {
                    symbol: "add1",
                },
            ),
        ),
    },
    edge weights: {
        0: 0,
        1: 0,
        2: 0,
    },
}
